### ISSUE TITLE
Function call support

### ISSUE THREAD

**Issue by georgemitenkov** [BlueBrain/nmodl#472]
Support function calls in NMODL
- locally defined functions
- mathematical built-ins

Question: can we call function external to the current mod file?
Question: does ordering of functions matter? (do we need declarations?)

**Comment by pramodk** [comment]
> can we call function external to the current mod file?

In theory yes but for what would like to do here, we can assume all functions are defined in the mod file. 

> does ordering of functions matter? (do we need declarations?)

IIRC, I read somewhere in language reference guide that order of method definitions doesn’t matter in the same modules. So prototype declaration is not required (?)

**Comment by georgemitenkov** [comment]
1) Cool, thanks! 
2) Here I mean if we are allowed to have this in NMODL
```
FUNCTION bar() {bar = foo()}
FUNCTION foo() {foo = 2}
```
Because traversing AST first meets `bar`, we need to create `call` instruction to `foo`. We pass `foo` it as a `llvm::Function*`, that in this case hasn't been created.

**Comment by pramodk** [comment]
> Here I mean if we are allowed to have this in NMODL
>   FUNCTION bar() {bar = foo()}
>   FUNCTION foo() {foo = 2}

Understood now! NMODL doesn't handle this i.e. order in the AST is same as how user write. For this reason, C++ code generation backend does emit prototype declarations for all functions and procedure at the beginning.

IIRC, prototype declaration is straightforward in LLVM IR, right? In that case it would be safe to emit them at the beginning. 

I suggest two options:
* if you see issues or complications to implement next option then assume that procedures or functions are in the "right" order and continue implementation.
* get list of functions + procedures and emit declarations at the beginning. For this, we can use `CodegenHelperVisitor`. As name suggest, this visitor  is used to traverse AST and "collect" necessary information to help code generation passes. It's used in https://github.com/BlueBrain/nmodl/blob/ab27595d047ef8e43519a5a6058d19f1a9bc45ef/src/codegen/codegen_c_visitor.cpp#L4375

Basically,  in `visit_program`, we can run `CodegenHelperVisitor` to populate [CodegenInfo](https://github.com/BlueBrain/nmodl/blob/ab27595d047ef8e43519a5a6058d19f1a9bc45ef/src/codegen/codegen_info.hpp#L139) structure which collect lot of information including list of functions and procedures:

```cpp
visit_program(const Program& node) {
    CodegenHelperVisitor v;
    auto& info = v.analyze(node);

    for (const auto& node: info.functions) {
        // emit declaration
    }
    for (const auto& node: info.procedures) {
         // emit declaration
    }
....
}
```

We do the same in C++ backend here https://github.com/BlueBrain/nmodl/blob/ab27595d047ef8e43519a5a6058d19f1a9bc45ef/src/codegen/codegen_c_visitor.cpp#L1416

**Comment by georgemitenkov** [comment]
Thanks! Yes, I just wanted to clarify before implementing these. I think that emitting declarations is a good and rather straightforward approach.
